dlt/destinations/impl/lancedb/lancedb_client.py

```diff
@@ -690,67 +691,177 @@ def complete_load(self, load_id: str) -> None:
+class LanceDBRemoveOrphansJob(RunnableLoadJob):
+    orphaned_ids: Set[str]
+
+    def __init__(
+        self,
+        file_path: str,
+    ) -> None:
+        super().__init__(file_path)
+        self._job_client: "LanceDBClient" = None
+        self.references = ReferenceFollowupJob.resolve_references(file_path)
+
+    def run(self) -> None:
+        db_client: DBConnection = self._job_client.db_client
+        id_field_name: str = self._job_client.config.id_field_name
+
+        # We don't all insert jobs for each table using this method.
+        table_lineage: List[TTableSchema] = []
+        for file_path_ in self.references:
+            table = self._schema.get_table(ParsedLoadJobFileName.parse(file_path_).table_name)
+            if table["name"] not in [table_["name"] for table_ in table_lineage]:
+                table_lineage.append(table)
+
+        for table in table_lineage:
+            fq_table_name: str = self._job_client.make_qualified_table_name(table["name"])
+            try:
+                fq_parent_table_name: str = self._job_client.make_qualified_table_name(
+                    table["parent"]
+                )
+            except KeyError:
+                fq_parent_table_name = None  # The table is a root table.
+
+            try:
+                child_table = db_client.open_table(fq_table_name)
+                child_table.checkout_latest()
+                if fq_parent_table_name:
+                    parent_table = db_client.open_table(fq_parent_table_name)
+                    parent_table.checkout_latest()
+            except FileNotFoundError as e:
+                raise DestinationTransientException(
+                    "Couldn't open lancedb database. Orphan removal WILL BE RETRIED"
+                ) from e
+
+            try:
+                if fq_parent_table_name:
+                    # Chunks and embeddings in child table.
+                    parent_ids = set(
+                        pc.unique(
+                            parent_table.to_lance().to_table(columns=["_dlt_id"])["_dlt_id"]
+                        ).to_pylist()
+                    )
+                    child_ids = set(
+                        pc.unique(
+                            child_table.to_lance().to_table(columns=["_dlt_parent_id"])[
+                                "_dlt_parent_id"
+                            ]
+                        ).to_pylist()
+                    )
+
+                    if orphaned_ids := child_ids - parent_ids:
+                        if len(orphaned_ids) > 1:
+                            child_table.delete(f"_dlt_parent_id IN {tuple(orphaned_ids)}")
+                        elif len(orphaned_ids) == 1:
+                            child_table.delete(f"_dlt_parent_id = '{orphaned_ids.pop()}'")
+
+                else:
+                    # Chunks and embeddings in the root table.
+                    document_id_field = get_columns_names_with_prop(table, DOCUMENT_ID_HINT)
```

Comment:
in adapter you mention that DOCUMENT ID HINT will be appended to primary key but here I see something else...

IMO what we want here
primary_key = (DOCUMENT ID, CHUNK ID)

so what adapter must make sure is: DOCUMENT ID is set on a subset of PRIMARY KEY

for example primary_key = ("doc_id", "chunk_id"), document_id=("doc_id",) leaving "chunk_id" as chunk_id

you can force both chunk and doc ids to be single columns if that simplifies your code

Reply:
Turns out that LanceDB currently does not have built-in support for primary keys: https://github.com/lancedb/lancedb/issues/1120
 
 The Lance merge methods also [claims to support compound merge keys with an Iterable[str]](https://lancedb.github.io/lance/api/python/lance.html#lance.dataset.LanceDataset.merge_insert)